flatland_plugins/src/laser.cpp

```diff
@@ -161,42 +161,55 @@ void Laser::ComputeLaserRanges() {
   v_world_laser_origin_ = m_world_to_laser_ * v_zero_point_;
 
   // Conver to Box2D data types
-  b2Vec2 laser_point;
   b2Vec2 laser_origin_point(v_world_laser_origin_(0), v_world_laser_origin_(1));
 
-  // loop through the laser points and call the Box2D world raycast
+  // Results vector
+  std::vector<std::future<std::pair<double, double>>> results(
+      laser_scan_.ranges.size());
+
+  // loop through the laser points and call the Box2D world raycast by
+  // enqueueing the callback
+  for (unsigned int i = 0; i < laser_scan_.ranges.size(); ++i) {
+    results[i] =
+        pool_.enqueue([i, this, laser_origin_point] {  // Lambda function
+          b2Vec2 laser_point;
+          laser_point.x = m_world_laser_points_(0, i);
+          laser_point.y = m_world_laser_points_(1, i);
+          LaserCallback cb(this);
+
+          GetModel()->GetPhysicsWorld()->RayCast(&cb, laser_origin_point,
+                                                 laser_point);
+
+          if (!cb.did_hit_) {
+            return std::make_pair<double, double>(NAN, 0);
+          } else {
+            return std::make_pair<double, double>(
+                cb.fraction_ * this->range_ + this->noise_gen_(this->rng_),
```

Comment:
std::normal_distribution might not be thread safe.  Is it possible to move this noise generator to cb object so that nobody shares it.

Reply:
I can move the noise out of the thread